server/src/main/java/org/elasticsearch/search/vectors/ChildBlockJoinVectorScorerProvider.java

```diff
@@ -0,0 +1,148 @@
+    }
+
+    private final Weight childrenFilter;
+    private final Weight nearestChildren;
+    private final RandomVectorScorerProvider scorerProvider;
+    private final BitSetProducer parentsFilter;
+    private final int numChildrenPerParent;
+
+    protected ChildBlockJoinVectorScorerProvider(
+        @Nullable Weight childrenFilter,
+        Weight nearestChildren,
+        RandomVectorScorerProvider scorerProvider,
+        BitSetProducer parentsFilter,
+        int numChildrenPerParent
+    ) {
+        this.childrenFilter = childrenFilter;
+        this.nearestChildren = nearestChildren;
+        this.scorerProvider = scorerProvider;
+        this.parentsFilter = parentsFilter;
+        this.numChildrenPerParent = numChildrenPerParent;
+    }
+
+    public ChildBlockJoinVectorScorer scorer(LeafReaderContext context) throws IOException {
+        Scorer originallyMatchedChildren = nearestChildren.scorer(context);
+        if (originallyMatchedChildren == null) {
+            return null;
+        }
+        final BitSet parentBitSet = parentsFilter.getBitSet(context);
+        if (parentBitSet == null) {
+            return null;
+        }
+        Scorer childFilterScorer = childrenFilter == null ? null : childrenFilter.scorer(context);
+        VectorScorer scorer = scorerProvider.get(context);
+        DocIdSetIterator childFilterIterator = childFilterScorer == null
+            ? DocIdSetIterator.all(context.reader().maxDoc())
+            : childFilterScorer.iterator();
+        DocIdSetIterator childVectorIterator = ConjunctionUtils.createConjunction(
+            List.of(childFilterIterator, scorer.iterator()),
+            List.of()
+        );
+
+        return new ChildBlockJoinVectorScorer(
+            context.docBase,
+            originallyMatchedChildren.iterator(),
+            childVectorIterator,
+            parentBitSet,
+            scorer,
+            numChildrenPerParent
+        );
+    }
+
+    static class ChildBlockJoinVectorScorer {
+        private final VectorScorer vectorScorer;
+        private final DocIdSetIterator previouslyFoundChildren;
+        private final DocIdSetIterator childFilterIterator;
+        private final BitSet parentBitSet;
+        private int currentParent = -1;
+        private final HitQueue queue;
+        private final int docBase;
+
+        protected ChildBlockJoinVectorScorer(
+            int docBase,
+            DocIdSetIterator previouslyFoundChildren,
+            DocIdSetIterator childFilterIterator,
+            BitSet parentBitSet,
+            VectorScorer vectorScorer,
+            int numChildrenPerParent
+        ) {
+            this.docBase = docBase;
+            this.vectorScorer = vectorScorer;
+            this.childFilterIterator = childFilterIterator;
+            this.previouslyFoundChildren = previouslyFoundChildren;
+            this.parentBitSet = parentBitSet;
+            this.queue = new HitQueue(numChildrenPerParent, false);
```

Comment:
You may ask "Ben, why not just score all and return all vectors because you gotta iterate all the children anyways to find the next nearest". Well, there are degenerate cases we want to avoid. What if one vector has 10k children?